internal/gw/client.go

```diff
@@ -145,40 +146,125 @@ func (c *client) uploadBlob(ctx context.Context, item walk.SyncItem) error {
 	return nil
 }
 
-func (c *client) EnsureUploaded(
+func (c *client) isDuplicate(
+	ctx context.Context,
+	checkItem walk.SyncItem,
+	items []walk.SyncItem,
+) bool {
+	for _, item := range items {
+		if item.Key == checkItem.Key || item.LinkTo == checkItem.Key {
+			log.FromContext(ctx).F("uri", item.SrcPath).Debug("Skipping duplicate item")
+			return true
+		}
+	}
+	return false
+}
+
+func (c *client) GetUploadItems(
 	ctx context.Context,
 	items []walk.SyncItem,
-	onUploaded func(walk.SyncItem) error,
 	onPresent func(walk.SyncItem) error,
-) error {
-	// TODO: concurrency
+	onDuplicate func(walk.SyncItem) error,
+) ([]walk.SyncItem, error) {
+	var toUpload []walk.SyncItem
 	for _, item := range items {
 		if item.Key == "" && item.LinkTo != "" {
 			log.FromContext(ctx).F("uri", item.SrcPath).Debug("Skipping unfollowed symlink")
 			continue
 		}
 
-		// Check if it's present
+		// Determine if the blob is already present in the bucket
 		have, err := c.haveBlob(ctx, item)
 		if err != nil {
-			return fmt.Errorf("checking for presence of %s: %w", item.Key, err)
+			return toUpload, fmt.Errorf("checking for presence of %s: %w", item.Key, err)
 		}
-
+		// If so, do not include it in the final set of items to upload
 		if have {
 			if err = onPresent(item); err != nil {
-				return err
+				return toUpload, err
 			}
 			continue
 		}
 
-		if err = c.uploadBlob(ctx, item); err != nil {
-			return err
+		// Determine if the item already exists in the final set of items to upload
+		have = c.isDuplicate(ctx, item, toUpload)
+		// If so, prevent a costly re-upload of the item
+		if have {
+			if err = onDuplicate(item); err != nil {
+				return toUpload, err
+			}
+			continue
 		}
-		if err = onUploaded(item); err != nil {
-			return err
+		toUpload = append(toUpload, item)
+	}
+	return toUpload, nil
+}
+
+type uploadResult struct {
+	Error error
+	Item  walk.SyncItem
+}
+
+func (c *client) uploadBlobs(
+	ctx context.Context,
+	items []walk.SyncItem,
+	ch chan uploadResult,
+	wg *sync.WaitGroup,
+) {
+	for _, item := range items {
+		if err := c.uploadBlob(ctx, item); err != nil {
+			ch <- uploadResult{err, item}
+			break
+		}
+		ch <- uploadResult{nil, item}
+	}
+	wg.Done()
+	return
+}
+
+func (c *client) EnsureUploaded(
+	ctx context.Context,
+	items []walk.SyncItem,
+	onUploaded func(walk.SyncItem) error,
+) error {
+	numThreads := c.cfg.GoroutineCount()
+	if numThreads > len(items) {
+		numThreads = len(items)
+	}
+	partitionSize := len(items) / numThreads
```

Comment:
I wouldn't say it's strictly necessary to change this to meet AC of the issue, but I don't think this has used the ideal or idiomatic way of splitting up the work across goroutines.

What you do here is split up the work into evenly sized chunks ahead of time, but the problem with this is you have no idea if the file sizes are uniformly distributed among the chunks.  If you imagine for example that you get unlucky and the very first chunk happens to contain a bunch of large ISOs, with other chunks containing much smaller files, then you quickly arrive at the situation where all goroutines other than the first have completed, and we are just waiting on the first to complete.  That brings us back to only uploading one thing at a time even though we've been requested to use more concurrency.

I think the right way is to use a channel to get data into the goroutines.  All the items for upload should be written into a channel; then each goroutine, rather than looping over an array, will loop over that channel until either the channel is closed (no more items), an error occurs, or the context is closed.  So the responsibility for picking up the next work item belongs to each goroutine and they will all stay busy until all the work is done.

This is how concurrency already works in some other parts of the code.  I think func getSyncItems in walk.go is the most applicable precedent: you see it makes one goroutine that writes items to a channel, and then spawns 20 (hardcoded) goroutines which all read items from the channel.  The way fillItems works there is how we'd want uploadBlobs to work here.